# An application for book lovers

<a id=cont></a>
## Table of contents

__[Project description](#desc)__

**1.** __[Download the data](#id1)__


**2.** __[Queries](#id2)__

**3.** __[Conclusions](#conc)__

<a id=desc></a>
## Project description

The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were at home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers. 

We've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

<a id=id1></a>
## 1. Download the data

Import the libraries

In [1]:
import pandas as pd
from sqlalchemy import create_engine
from simple_colors import *

**(!)** The tables are in a PostgreSQL database with the following structure:

![](https://github.com/EkaterinaNik2021/Book_Lovers_App/blob/main/1.png)

**(!)** Defining parameters for connecting to the database on PostgreSQL.

In [2]:
#parameters
db_config = {'user': 'practicum_student',         # username
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the database
#database connection string
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])
#connecting to the database
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

**(!)** Let's load the data and do an initial inspection.

In [3]:
query = ''' SELECT * FROM books '''
books_df =pd.io.sql.read_sql(query, con = engine)

query_1 = ''' SELECT * FROM authors '''
authors_df =pd.io.sql.read_sql(query_1, con = engine)

query_2 = ''' SELECT * FROM publishers '''
publishers_df =pd.io.sql.read_sql(query_2, con = engine)

query_3 = ''' SELECT * FROM ratings '''
ratings_df =pd.io.sql.read_sql(query_3, con = engine)

query_4 = ''' SELECT * FROM reviews '''
reviews_df =pd.io.sql.read_sql(query_4, con = engine)

books_df.name='BOOKS'
authors_df.name='AUTHORS'
publishers_df.name='PUBLISHERS'
ratings_df.name='RATINGS'
reviews_df.name='REVIEWS'


def get_info(df):
    
    print('Some information about table: '+ black(df.name, 'bold'))
    print()
    print('INFO:')
    print()
    display(df.info(memory_usage='deep'))
    print()
    print('**********'*10)
    print()
    print('DESCRIBE:')
    display(df.describe(include='all'))
    print()
    print('**********'*10)
    print()
    print('FIRST 10 ROWS:')
    display(df.head(10))
    print()
    print('**********'*10)
    
get_info(books_df)
get_info(authors_df)
get_info(publishers_df)
get_info(ratings_df)
get_info(reviews_df)


Some information about table: BOOKS

INFO:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 156.3 KB


None


****************************************************************************************************

DESCRIBE:


book_id    author_id                title   num_pages  \
count   1000.000000  1000.000000                 1000  1000.00000   
unique          NaN          NaN                  999         NaN   
top             NaN          NaN  Memoirs of a Geisha         NaN   
freq            NaN          NaN                    2         NaN   
mean     500.500000   320.417000                  NaN   389.11100   
std      288.819436   181.620172                  NaN   229.39014   
min        1.000000     1.000000                  NaN    14.00000   
25%      250.750000   162.750000                  NaN   249.00000   
50%      500.500000   316.500000                  NaN   352.00000   
75%      750.250000   481.000000                  NaN   453.00000   
max     1000.000000   636.000000                  NaN  2690.00000   

       publication_date  publisher_id  
count              1000   1000.000000  
unique              618           NaN  
top          2004-06-01           NaN  
freq                 10           NaN  
mean                NaN    171.270000  
std                 NaN     99.082685  
min                 NaN      1.000000  
25%                 NaN     83.000000  
50%                 NaN    177.500000  
75%                 NaN    258.000000  
max                 NaN    340.000000


****************************************************************************************************

FIRST 10 ROWS:


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116


****************************************************************************************************
Some information about table: AUTHORS

INFO:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 56.5 KB


None


****************************************************************************************************

DESCRIBE:


author_id         author
count   636.000000            636
unique         NaN            636
top            NaN  Jen Lancaster
freq           NaN              1
mean    318.500000            NaN
std     183.741666            NaN
min       1.000000            NaN
25%     159.750000            NaN
50%     318.500000            NaN
75%     477.250000            NaN
max     636.000000            NaN


****************************************************************************************************

FIRST 10 ROWS:


author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts


****************************************************************************************************
Some information about table: PUBLISHERS

INFO:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 27.2 KB


None


****************************************************************************************************

DESCRIBE:


publisher_id      publisher
count     340.000000            340
unique           NaN            340
top              NaN  Thomas Nelson
freq             NaN              1
mean      170.500000            NaN
std        98.293777            NaN
min         1.000000            NaN
25%        85.750000            NaN
50%       170.500000            NaN
75%       255.250000            NaN
max       340.000000            NaN


****************************************************************************************************

FIRST 10 ROWS:


publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers


****************************************************************************************************
Some information about table: RATINGS

INFO:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 571.1 KB


None


****************************************************************************************************

DESCRIBE:


rating_id      book_id    username       rating
count   6456.000000  6456.000000        6456  6456.000000
unique          NaN          NaN         160          NaN
top             NaN          NaN  martinadam          NaN
freq            NaN          NaN          56          NaN
mean    3228.500000   510.574195         NaN     3.928284
std     1863.831001   284.141636         NaN     0.943303
min        1.000000     1.000000         NaN     1.000000
25%     1614.750000   291.000000         NaN     3.000000
50%     3228.500000   506.000000         NaN     4.000000
75%     4842.250000   750.000000         NaN     5.000000
max     6456.000000  1000.000000         NaN     5.000000


****************************************************************************************************

FIRST 10 ROWS:


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4


****************************************************************************************************
Some information about table: REVIEWS

INFO:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 630.3 KB


None


****************************************************************************************************

DESCRIBE:


review_id      book_id username  \
count   2793.000000  2793.000000     2793   
unique          NaN          NaN      160   
top             NaN          NaN  susan85   
freq            NaN          NaN       29   
mean    1397.000000   504.693161      NaN   
std      806.413976   288.472931      NaN   
min        1.000000     1.000000      NaN   
25%      699.000000   259.000000      NaN   
50%     1397.000000   505.000000      NaN   
75%     2095.000000   753.000000      NaN   
max     2793.000000  1000.000000      NaN   

                                                     text  
count                                                2793  
unique                                               2793  
top     Than throughout occur decision right. Woman ri...  
freq                                                    1  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN


****************************************************************************************************

FIRST 10 ROWS:


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...


****************************************************************************************************


__[Back to content](#cont)__

<a id=id2></a>
## 2. Queries

***Task 1.*** Find the number of books released after January 1, 2000.

In [4]:
query = ''' SELECT COUNT(book_id) as book_num 
            FROM books 
            WHERE CAST(publication_date AS date) > '2000-01-01'
         '''
books_num=pd.io.sql.read_sql(query, con = engine)
print('The number of books released after January 1,2000: '+ blue(books_num['book_num'].iloc[0],'bold'))


The number of books released after January 1,2000: 819


***Task 2.*** Find the number of user reviews and the average rating for each book.

In [5]:
query= ''' SELECT 
             books.book_id, 
             books.title, 
             COUNT(DISTINCT reviews.review_id) as review_count,
             AVG(ratings.rating) as rating_avg
           FROM books
           LEFT JOIN reviews ON books.book_id=reviews.book_id
           LEFT JOIN ratings ON books.book_id=ratings.book_id
           GROUP BY books.book_id
           ORDER BY review_count DESC
        '''
books_rating=pd.io.sql.read_sql(query, con = engine)
books_rating

book_id                                              title  review_count  \
0        948                            Twilight (Twilight  #1)             7   
1        963                                Water for Elephants             6   
2        734                                   The Glass Castle             6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...             6   
4        695  The Curious Incident of the Dog in the Night-Time             6   
..       ...                                                ...           ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
996      808                            The Natural Way to Draw             0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites             0   
998      221                          Essential Tales and Poems             0   
999      191  Disney's Beauty and the Beast (A Little Golden...             0   

     rating_avg  
0      3.662500  
1      3.977273  
2      4.206897  
3      4.414634  
4      4.081081  
..          ...  
995    3.666667  
996    3.000000  
997    5.000000  
998    4.000000  
999    4.000000  

[1000 rows x 4 columns]

***Task 3.*** Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).

In [6]:
query = ''' SELECT
              publishers.publisher,
              COUNT(books.book_id) as books_num
            FROM publishers
            LEFT JOIN books ON publishers.publisher_id=books.publisher_id
            WHERE books.num_pages>50
            GROUP BY publishers.publisher
            ORDER BY books_num DESC
            LIMIT 1
        '''
publisher=pd.io.sql.read_sql(query, con = engine)
print('The publisher with the greatest number of released books with more than 50 pages: '\
      + blue(publisher['publisher'].iloc[0], 'bold'))

The publisher with the greatest number of released books with more than 50 pages: Penguin Books


***Task 4.*** Identify the author with the highest average book rating (look only at books with at least 50 ratings).

In [7]:
query= ''' SELECT 
             authors.author,
             books.book_id,
             COUNT(ratings.rating_id) as rating_count,
             AVG(ratings.rating) as rating_avg
           FROM authors
           LEFT JOIN books ON authors.author_id=books.author_id
           LEFT JOIN ratings ON ratings.book_id=books.book_id
           GROUP BY authors.author, books.book_id
           HAVING COUNT(ratings.rating_id)>50
           ORDER BY rating_avg DESC
           LIMIT 1
       '''
author=pd.io.sql.read_sql(query, con = engine)
print('The author with the highest average book rating: ' +blue(author['author'].iloc[0], 'bold'))

The author with the highest average book rating: J.K. Rowling/Mary GrandPré


***Task 5.*** Find the average number of text reviews among users who rated more than 50 books.

In [8]:
query= ''' SELECT AVG(SUBQ.num_text) AS avg_num_text
           FROM (
                 SELECT COUNT(text) AS num_text
                 FROM reviews
                 WHERE username IN
                   ( SELECT username
                     FROM ratings
                     GROUP BY username
                     HAVING COUNT(DISTINCT book_id)>50
                     ORDER BY COUNT(DISTINCT book_id) DESC )
                GROUP BY username ) AS SUBQ
          
       '''
rat_avg=pd.io.sql.read_sql(query, con = engine)
print('The average number of text reviews among users who rated more than 50 books is: '+ blue(round(rat_avg['avg_num_text'].iloc[0],2),'bold'))

The average number of text reviews among users who rated more than 50 books is: 24.33


__[Back to content](#cont)__

<a id=id3></a>
## 3. Conclusions

**1. A few words about data:**

1. Our Dataset consists of 5 tables.
2. There are **999** unique books.
3. There are **636** authors.
4. There are **340** publishers.
5. **6456** ratings from **160** unique users; average rating is **3.9**.
6. **2793** reviews from **160** unique users.

**Queries:**

***Task 1.*** Find the number of books released after January 1, 2000.

 * **A:** *The number of books released after January 1,2000: **819**.*


***Task 2.*** Find the number of user reviews and the average rating for each book.

 * **A:** *See the table above.*


***Task 3.*** Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).

 * **A:** *The publisher with the greatest number of released books with more than 50 pages: **Penguin Books**.*


***Task 4.*** Identify the author with the highest average book rating (look only at books with at least 50 ratings).

 * **A:** *The author with the highest average book rating: **J.K. Rowling/Mary GrandPré**.*


***Task 5.*** Find the average number of text reviews among users who rated more than 50 books.

 * **A:** *The average number of text reviews among users who rated more than 50 books is: **24.33**.*

__[Back to content](#cont)__